File: aat-android/src/main/java/ch/bailu/aat/services/location/Hypsometric.java
package ch.bailu.aat.services.location;

public final class Hypsometric {
    /**
     *  See
     *  https://physics.stackexchange.com/questions/333475/how-to-calculate-altitude-from-current-temperature-and-pressure
     */

    private final static double MIN_PRESSURE=100d;

    private final static double TEMPERATURE_CELSIUS = 15d;
    private final static double TEMPERATURE_KELVIN = TEMPERATURE_CELSIUS + 273.15d;
    private final static double Ex = 5.257d;
    private final static double Rx = 1d / Ex;
    private final static double Ax = 0.0065d;



    private double pressureAtSeaLevel = 0;
    private double pressure = 0;


    public void setPressure(double d) {
        pressure = d;
    }


    public void setAltitude(double a) {
        if (isPressureValid()) {
            pressureAtSeaLevel = getPressureAtSeaLevel(pressure, a);
        }
    }


    public double getAltitude() {
        if (isValid()) {
            return  (getAltitude(pressure, pressureAtSeaLevel));
        }

        return 0;
    }


    public double getPressureAtSeaLevel() {
        return pressureAtSeaLevel;
    }


    public boolean isPressureValid() {
        return pressure > MIN_PRESSURE;
    }

    public boolean isPressureAtSeaLevelValid() {
        return pressureAtSeaLevel > MIN_PRESSURE;
    }


    public boolean isValid() {
        return isPressureValid() && isPressureAtSeaLevelValid();
    }




    public static double getAltitude(double pressure, double pressureAtSeaLevel) {

        final double Px = pressureAtSeaLevel / pressure;

        return ((Math.pow(Px, Rx) -1d) * TEMPERATURE_KELVIN)/Ax;
    }


    public static double getPressureAtSeaLevel(double pressure, double altitude) {

        final double Hx = altitude * Ax;

        return pressure * Math.pow( (Hx / TEMPERATURE_KELVIN) + 1d, Ex);
    }

    public void setPressureAtSeaLevel(double pressure) {
        pressureAtSeaLevel = pressure;
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/sensor/internal/Hypsometric.kt
package ch.bailu.aat.services.sensor.internal

import kotlin.math.pow

class Hypsometric {
    var pressureAtSeaLevel = 0.0
    private var pressure = 0.0

    fun setPressure(d: Double) {
        pressure = d
    }

    var altitude: Double
        get() = if (isValid) {
            getAltitude(pressure, pressureAtSeaLevel)
        } else 0.0

        set(a) {
            if (isPressureValid) {
                pressureAtSeaLevel = getPressureAtSeaLevel(pressure, a)
            }
        }

    private val isPressureValid: Boolean
        get() = pressure > MIN_PRESSURE

    val isPressureAtSeaLevelValid: Boolean
        get() = pressureAtSeaLevel > MIN_PRESSURE

    private val isValid: Boolean
        get() = isPressureValid && isPressureAtSeaLevelValid

    companion object {
        /**
         * See
         * https://physics.stackexchange.com/questions/333475/how-to-calculate-altitude-from-current-temperature-and-pressure
         */
        private const val MIN_PRESSURE = 100.0
        private const val TEMPERATURE_CELSIUS = 15.0
        private const val TEMPERATURE_KELVIN = TEMPERATURE_CELSIUS + 273.15
        private const val Ex = 5.257
        private const val Rx = 1.0 / Ex
        private const val Ax = 0.0065
        fun getAltitude(pressure: Double, pressureAtSeaLevel: Double): Double {
            val px = pressureAtSeaLevel / pressure
            return (px.pow(Rx) - 1.0) * TEMPERATURE_KELVIN / Ax
        }

        fun getPressureAtSeaLevel(pressure: Double, altitude: Double): Double {
            val hx = altitude * Ax
            return pressure * (hx / TEMPERATURE_KELVIN + 1.0).pow(Ex)
        }
    }
}
